src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/AbstractWorkspaceDocumentDiagnosticSource.cs

```diff
@@ -40,14 +51,32 @@ public override bool IsLiveSource()
             }
             else
             {
-                // We call GetDiagnosticsForIdsAsync as we want to ensure we get the full set of diagnostics for this document
-                // including those reported as a compilation end diagnostic.  These are not included in document pull (uses GetDiagnosticsForSpan) due to cost.
-                // However we can include them as a part of workspace pull when FSA is on.
-                var documentDiagnostics = await diagnosticAnalyzerService.GetDiagnosticsForIdsAsync(
-                    Document.Project.Solution, Document.Project.Id, Document.Id,
-                    diagnosticIds: null, shouldIncludeAnalyzer, includeSuppressedDiagnostics: false,
-                    includeLocalDocumentDiagnostics: true, includeNonLocalDocumentDiagnostics: true, cancellationToken).ConfigureAwait(false);
-                return documentDiagnostics;
+                var projectDiagnostics = await GetProjectDiagnosticsAsync(diagnosticAnalyzerService, cancellationToken).ConfigureAwait(false);
+                return projectDiagnostics.WhereAsArray(d => d.DocumentId == Document.Id);
+            }
+        }
+
+        private async ValueTask<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(
+            IDiagnosticAnalyzerService diagnosticAnalyzerService, CancellationToken cancellationToken)
+        {
+            if (!s_projectToDiagnostics.TryGetValue(Document.Project, out var lazyDiagnostics))
```

Comment:
Might be good to add a comment on `s_projectToDiagnostics` that parallel access won't happen because
1.  There is only one workspace diagnostic bucket that creates the `AbstractWorkspaceDocumentDiagnosticSource`
2.  the sources are called in order

Reply:
Discussed offline.  It's correct that parallel access won't happen.  But it's also 100% ok if it does.  CWT (and the pattern we use to access it here) is concurrency safe.